ocaml/xapi/repository_helpers.ml

```diff
@@ -653,15 +653,92 @@ let eval_guidances ~updates_info ~updates ~kind ~livepatches ~failed_livepatches
   |> GuidanceSet.resort_guidances ~remove_evacuations:(kind = Guidance.Absolute)
   |> GuidanceSet.elements
 
-let merge_with_unapplied_guidances ~__context ~host ~guidances =
+let merge_with_unapplied_guidances ~unapplied ~coming =
+  let open Guidance in
   let open GuidanceSet in
-  Db.Host.get_pending_guidances ~__context ~self:host
-  |> List.map (fun g -> Guidance.of_update_guidance g)
-  |> List.filter (fun g -> g <> Guidance.RebootHostOnLivePatchFailure)
-  |> of_list
-  |> union (of_list guidances)
-  |> resort_guidances ~remove_evacuations:false
-  |> elements
+  let unapplied_set = of_list unapplied in
+  let coming_set = of_list coming in
+  iter (fun g -> debug "unapplied: %s" (to_string g)) unapplied_set ;
+  iter (fun g -> debug "coming: %s" (to_string g)) coming_set ;
+  let resorted =
+    union unapplied_set coming_set |> resort_guidances ~remove_evacuations:false
+  in
+  let unapplied_only = diff unapplied_set coming_set in
+  let to_be_added', to_be_removed', unchanged' =
+    match (is_empty unapplied_only, subset unapplied_only resorted) with
+    | true, _ ->
+        (resorted, empty, empty)
+    | false, true ->
+        (* The unapplied_only is contained in resorted *)
+        (diff resorted unapplied_only, empty, unapplied_only)
+    | false, false ->
+        (diff resorted unapplied_only, unapplied_only, empty)
```

Comment:
Can we get the result directly like (tried this change with the added UT, passed):

  let to_be_added' = diff resorted unapplied_only in
  let to_be_removed' = diff unapplied_set resorted in
  let kept_only_from_unapplied' = diff resorted coming_set in
  let to_be_added, to_be_removed, kept_untouched =
    (elements to_be_added', elements to_be_removed', elements kept_only_from_unapplied')
  in

  # might worth adding comment for the "kept_untouched"(or "unchanged") to make it more clear: those guidance which come only from "unapplied", and not from "coming", and after "resort", those guidance are still needed.